src/dd_comm.F90

```diff
@@ -0,0 +1,1104 @@
+
+    ! We're going to send sites one at a time, then ALLREDUCE the number of
+    ! sent sites so each rank knows how many revcs to post
+    allocate(n_send_rank(n_procs))
+    n_send_rank = 0
+
+    ! We need a copy of the send data
+    allocate(buffer(work), STAT=alloc_err)
+
+    ! Check for allocation errors
+    if (alloc_err /= 0) then
+      call fatal_error("Failed to allocate source bank send buffer during &
+           &domain decomposition initialization.")
+    end if
+
+    ! TODO: This may be too conservative an allocation for most use-cases
+    allocate(request(work + n_particles), STAT=alloc_err)
+    n_request = 0
+
+    ! Check for allocation errors
+    if (alloc_err /= 0) then
+      call fatal_error("Failed to allocate communication request array during &
+           &domain decomposition initialization.")
+    end if
+
+    ! Copy initially-sampled source sites
+    buffer = source_bank
+
+    ! We'll save sites that belong in this domain to the source bank as we
+    ! loop through and find them, as well as when we receive them from other
+    ! processes
+    n_source_sites = 0
+
+    ! Loop over all sampled source sites and send/save sites as needed
+    do i = 1, work
+
+      ! Determine which domain meshbin this should start on
+      call get_mesh_bin(dd % mesh, buffer(i) % xyz, to_domain)
+
+      if (to_domain == dd % meshbin) then
+        ! This site should start on this domain
+
+        n_source_sites = n_source_sites + 1
+        source_bank(n_source_sites) = buffer(i)
+
+      else
+        ! This site needs to be sent
+
+        ! We'll pick a process at random on that domain to send it to, using
+        ! the seed stored in the site for random number reproducibility
+        prn_seed = buffer(i) % prn_seed
+        to_rank = dd % domain_masters(to_domain) + &
+             int(prn() * real(dd % domain_n_procs(to_domain), 8))
+
+        ! Increment send counters
+        n_send_rank(to_rank + 1) = n_send_rank(to_rank + 1) + 1
+        n_send_domain(to_domain) = n_send_domain(to_domain) + 1
+
+        ! Post the send
+#ifdef MPI
+        n_request = n_request + 1
+        call MPI_ISEND(buffer(i), 1, MPI_BANK, to_rank, rank, MPI_COMM_WORLD, &
+             request(n_request), mpi_err)
+#endif
+      end if
+
+    end do
+
+    ! Determine how many sites were sent here by reducing n_send_rank
+#ifdef MPI
+    call MPI_ALLREDUCE(MPI_IN_PLACE, n_send_rank, n_procs, MPI_INTEGER, &
+         MPI_SUM, MPI_COMM_WORLD, mpi_err)
+#endif
+
+    ! Check if we're not going to have enough space in the source_bank (which
+    ! we're using as the receive buffer), and resize the array if needed
+    new_source = n_send_rank(rank + 1) + n_source_sites
+    if (new_source > size_source_bank) &
+         call extend_array(source_bank, size_source_bank, int(new_source, 8), &
+             alloc_err)
+        call resize_array(dd % particle_buffer, dd % size_particle_buffer, &
+             int(new_source, 8), alloc_err)
```

Comment:
I don't like the `resize_array` routine.  It's not clear from the name that it can only make arrays larger and that it erases the contents.  I think it's also strange that it requires and then modifies the `current_size` argument.  Please replace all occurrences of `resize_array` with `deallocate` and `allocate` statements, and delete its definition in resize_arr.F90